lib/web_catalog/api_extractor.es6.js

```diff
@@ -0,0 +1,279 @@
+];
+
+var blacklistFunctionAttr = ['prototype'];
+// blacklistFunctionAttr is list of property of Function.prototype,
+// prototype is not a part of function, but we also wish to blacklist it.
+
+var blacklistObjectAttr = [];
+// blacklistObjectAttr is list of property of Object.prototype.
+
+var constType = ["boolean", "number", "string"];
+// Primitive types which is constant will be blacklisted.
+
+var rootId, ObjectId, FunctionId, objectId, functionId;
+
+// A wrapper function of og.getObjectKeys, constant will
+// be filtered if showConst is not turned on.
+function getObjectProperties(og, id) {
+  var keys = og.getObjectKeys(id);
+  keys = keys.filter(key => {
+    var keyId = og.lookup(key, id);
+    var meta = og.lookupMetaData(key, id);
+    if (Object.keys(meta).length === 0) {
+      // If meta contains no information, it means this property
+      // is inherited from its prototype. Inherited properties
+      // will be added to thier prototype interface.
+      return false;
+    }
+    if (!showConst) {
+      // If showConst is not turned on, filter the result if
+      // string it is premetive type and writable is false.
+      return constType.indexOf(og.getType(keyId)) === -1 || meta.writable !== 0;
+    }
+    return true;
+  });
+  return keys;
+}
+
+// Remove all +, $ chars in given string, return the new string.
+function cleanUp(str) {
+  return str.replace(/\+/g, '').replace(/\$/g, '');
+}
+
+// Return if str is a positive interger.
+function isPosotiveInt(str) {
+  return /^\+?(0|[1-9]\d*)$/.test(str);
+}
+
+// Merge all given arrays into arr, no dupulicate allowed,
+// blacklisted properties are filtered.
+function arrayMerge(arr, opt) {
+  opt = opt || {};
+
+  for (var i = 1; i < arguments.length; i += 1) {
+    for (var j = 0; j < arguments[i].length; j += 1) {
+      var str = cleanUp(arguments[i][j]);
+      if (arr.indexOf(str) === -1 &&
+        ((blacklistObjectAttr.indexOf(str) === -1 &&
+        blacklistFunctionAttr.indexOf(str) === -1) || noise) &&
+        !isPosotiveInt(str)) {
+        arr.push(str);
+      }
+    }
+  }
+}
+
+// Return an array which is arr1 set minus arr2.
+function setMinus(arr1, arr2) {
+  var retArr = [];
+  for (let i = 0; i < arr1.length; i += 1) {
+    var str = cleanUp(arr1[i]);
+    if (arr2.indexOf(str) === -1) {
+      // Add str to restul Array if cannot find str in arr2.
+      retArr.push(str);
+    }
+  }
+  return retArr;
+}
+
+// Get list of property from Function.prototype
+// and push to blacklistFunctionAttr;
+function getFunctionBlacklist(og) {
+  blacklistFunctionAttr = [];
+  FunctionId = og.lookup('Function', rootId);
+  functionId = og.lookup('prototype', FunctionId);
+  var functionProp = og.getObjectKeys(functionId);
+  for (var i = 0; i < functionProp.length; i += 1) {
+    blacklistFunctionAttr.push(cleanUp(functionProp[i]));
+  }
+}
+
+// Get list of property from Object.prototype
+// and push to blacklistObjectAttr;
+function getObjectBlacklist(og) {
```

Comment:
Is this the same algorithm as `getFunctionBlacklist`? Maybe we could have a common `getPrototypePropertyProperties` that looks up the property named `prototype` on some given lookup-name or id, then returns an array of its properties.